File: code/ex710.java
class Solution {
    public Map<Integer, Integer> reflection = new HashMap<>();
    public Random rand = new Random();
    public int frontLen;

    public Solution(int N, int[] blacklist) {
        int n = blacklist.length;
        frontLen = N - n;
        
        // Seperate [0, N) to two parts: good part & black points
        // Get the good points in the second part
        Set<Integer> set = new HashSet<>();
        for (int i = frontLen; i < N; i++)    set.add(i);
        for (int b : blacklist)    set.remove(b);

        // Reflect good points in the second list to the position of bad points in the first list
        Iterator<Integer> it = set.iterator();
        for (int b : blacklist) {
            if (b < frontLen) {
                reflection.put(b, it.next());
            }
        }
    }
    
    public int pick() {
        int r = rand.nextInt(frontLen);
        return reflection.getOrDefault(r, r);
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(N, blacklist);
 * int param_1 = obj.pick();
 */

File: code/ex710.cpp
class Solution {
    unordered_map<int, int> mirror;
    int n;
public:
    Solution(int N, vector<int>& blacklist) {
        unordered_set<int> frontbl(blacklist.begin(), blacklist.end());
        int frontlen = N - blacklist.size();
        n = frontlen;
        // Delete black holes outside the front range
        for (int i = frontlen; i < N; i++) {
            if (frontbl.find(i) != frontbl.end())    frontbl.erase(i);
        }
        
        // Mirror white elements outside front range to black holes with in front range
        unordered_set<int> bl(blacklist.begin(), blacklist.end());
        auto it = frontbl.begin();
        for (int i = frontlen; i < N; i++) {
            if (bl.find(i) != bl.end())    continue;    // Skip black holes
            mirror[*it] = i;
            it++;
        }
    }
    
    int pick() {
        int random = rand() % n;
        return mirror.find(random) != mirror.end() ? mirror[random] : random;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(N, blacklist);
 * int param_1 = obj->pick();
 */